qdrant-landing/content/blog/colpali-optimization.md

```diff
@@ -54,54 +61,52 @@ For those eager to explore, the [codebase is available here](https://github.com/
 
 Pooling is well-known in machine learning as a way to compress data while keeping important information intact. For ColPali, we reduced ~1,030 vectors per page to just 38 vectors by pooling rows in the document's 32x32 grid.
 
-The most popular types of pooling are:
+Max and mean pooling are the two most popular types, so we decided to test both approaches on the rows of the grid. Likewise, we could apply pooling on columns, which we plan to explore in the future.
 
 - **Mean Pooling:** Averages values across rows.
 - **Max Pooling:** Selects the maximum value for each feature.
 
-32 vectors represent the pooled rows, while an additional 6 vectors encode contextual information derived from ColPali’s special tokens (e.g., <bos> for the beginning of the sequence, and task-specific instructions like “Describe the image”).
+32 vectors represent the pooled rows, while the final 6 vectors encode contextual information derived from ColPali’s special tokens (e.g., <bos> for the beginning of the sequence, and task-specific instructions like “Describe the image”).
 
 For our experiments, we chose to preserve these 6 additional vectors.
 
 ### The "ColPali as a Reranker" Experiment
 
-Pooling drastically reduces retrieval costs, but there’s a risk of losing fine-grained precision. To address this, we implemented a **two-stage retrieval system** with the pooled vectors for the initial retrieval stage and the original ColPali model for reranking:
+Pooling drastically reduces retrieval costs, but there’s a risk of losing fine-grained precision. To address this, we implemented a **two-stage retrieval system**, where embeddings generated with ColPali were max/mean pooled by grid rows to create lightweight vectors for the initial retrieval stage, followed by reranking with the original high-resolution embeddings:
 
 1. **Pooled Retrieval:** Quickly retrieves the top 200 candidates using lightweight pooled embeddings.
 2. **Full Reranking:** Refines these candidates using the original, high-resolution embeddings, delivering the final top 20 results.
 
-This approach delivers speed without compromising on retrieval quality.
-
 ### Implementation
 
 We created a custom dataset with over 20,000 unique PDF pages by merging:
 
-- **ViDoRe Benchmark:** Designed for document retrieval evaluation.
+- **ViDoRe Benchmark:** Designed for PDF documents retrieval evaluation.
 - **UFO Dataset:** Visually rich documents paired with synthetic queries.
 - **DocVQA Dataset:** A large set of document-derived Q&A pairs.
 
 Each document was processed into 32x32 grids, generating both full-resolution and pooled embeddings. 
 
 ![](/blog/colpali-optimization/rows.png)
```

Comment:
I feel like maybe image should come earlier, where we talk about pooling first time